packages/scripts/src/upgrade-tests/162/1_Propose.ts

```diff
@@ -1,85 +1,344 @@
+      isPolygon ? "GovernorChildTunnel" : "GovernorSpoke",
+      l2NodeUrl
     );
+    const l2Finder = await getContractInstanceByUrl<Finder>("Finder", l2NodeUrl);
+
+    if (await l2Registry.isContractRegistered(l2OptimisticOracleV2Address)) continue;
+
+    console.log(`Registering ${l2OptimisticOracleV2Address} on ${networkName}`);
+
+    // Fund Arbitrum if needed for next 4 transactions
+    if (isArbitrum) await fundArbitrumParentMessengerForRelays(arbitrumParentMessenger, proposerSigner, 4);
+
+    // 1. Temporarily add the GovernorChildTunnel/GovernorSpoke  as a contract creator.
+    const addMemberDataTx = await l2Registry.populateTransaction.addMember(
+      RegistryRolesEnum.CONTRACT_CREATOR,
+      l2Governor.address
+    );
+    adminProposalTransactions.push(
+      await relayGovernanceMessage(
+        l2Registry.address,
+        addMemberDataTx,
+        isPolygon ? governorRootTunnel : governorHub,
+        l2NetworkId,
+        relayRecords
+      )
+    );
+
+    console.log("- addMemberData", addMemberDataTx);
+
+    // 2. Register the OptimisticOracle as a verified contract.
+    const registerOptimisticOracleData = await l2Registry.populateTransaction.registerContract(
+      [],
+      l2OptimisticOracleV2Address
+    );
+
+    adminProposalTransactions.push(
+      await relayGovernanceMessage(
+        l2Registry.address,
+        registerOptimisticOracleData,
+        isPolygon ? governorRootTunnel : governorHub,
+        l2NetworkId,
+        relayRecords
+      )
+    );
+
+    console.log("registerOptimisticOracleData", registerOptimisticOracleData);
+
+    // 3. Remove the l2Governor from being a contract creator.
+    const removeMemberData = await l2Registry.populateTransaction.removeMember(
+      RegistryRolesEnum.CONTRACT_CREATOR,
+      l2Governor.address
+    );
+
+    adminProposalTransactions.push(
+      await relayGovernanceMessage(
+        l2Registry.address,
+        removeMemberData,
+        isPolygon ? governorRootTunnel : governorHub,
+        l2NetworkId,
+        relayRecords
+      )
+    );
+
+    console.log("- removeMemberData", removeMemberData);
+
+    // 4. Set contract in finder.
+    const setFinderData = await l2Finder.populateTransaction.changeImplementationAddress(
+      hre.ethers.utils.formatBytes32String(OPTIMISTIC_ORACLE_V2),
+      l2OptimisticOracleV2Address
+    );
+
+    adminProposalTransactions.push(
+      await relayGovernanceMessage(
+        l2Finder.address,
+        setFinderData,
+        isPolygon ? governorRootTunnel : governorHub,
+        l2NetworkId,
+        relayRecords
+      )
+    );
+
+    console.log("- changeImplementationAddressData", setFinderData);
+  }
+
+  // store relayRecords
+  const relayRecordsJson = JSON.stringify(relayRecords, null, 2);
+  const relayRecordsFile = path.join(path.dirname(__filename), "response.json");
+
+  // delete file if exists
+  if (fs.existsSync(relayRecordsFile)) {
+    fs.unlinkSync(relayRecordsFile);
```

Comment:
I'm a little confused -- why do we need to save a file?

Reply:
I was putting the call data in a JSON file so that the verification script could use it. It was not a very good idea.

I have changed the verification to use the decodeData script.